## Actividad V- Conexión a BD

**Consigna**

1. Deberán generar una instancia de conexión a Base de datos preferiblemente PostgreSQL donde deberá existir el dataset elegido para el proyecto.

2. Se deben realizar consultas SQL desde el lenguaje Python utilizando las librerías (psycopg2 o sqlalchemy) y quedar evidenciadas en el archivo .ipynb que se entregue.

**Desarrollo**

1. Me interesa conocer el comportamiento de la popularidad de ventas de vehículos eléctricos, en éste caso se evalúan datos del Estado de Washington en Estados Unidos.

Seleccioné el siguiente Dataset para la actividad.
Este Dataset muestra los Vehículos de Batería Eléctrica (BEVs) y Vehículos Eléctricos Híbridos (PHEVs), registrados actualmente en el Departamento de Licencias del Estado de Washington (DOL). Metadata Actualizada: March 16, 2024 URL: https://catalog.data.gov/dataset/electric-vehicle-population-data




In [1]:
import pandas as pd
import numpy as np
from sqlalchemy import create_engine
import psycopg2

**Leemos datos con Pandas primero para compararlo después cuando se carguen mediante la base de datos en PostgreSQL**

In [2]:
Dataset = pd.read_csv("https://raw.githubusercontent.com/Fab-RG/Bootcamp-DS/main/Actividad%202/Electric_Vehicle_Population_Data.csv", sep=",")
Dataset.head()

VIN (1-10)     County       City State  Postal Code  Model Year   Make  \
0  5YJYGDEE1L       King    Seattle    WA      98122.0        2020  TESLA   
1  7SAYGDEE9P  Snohomish    Bothell    WA      98021.0        2023  TESLA   
2  5YJSA1E4XK       King    Seattle    WA      98109.0        2019  TESLA   
3  5YJSA1E27G       King   Issaquah    WA      98027.0        2016  TESLA   
4  5YJYGDEE5M     Kitsap  Suquamish    WA      98392.0        2021  TESLA   

     Model           Electric Vehicle Type  \
0  MODEL Y  Battery Electric Vehicle (BEV)   
1  MODEL Y  Battery Electric Vehicle (BEV)   
2  MODEL S  Battery Electric Vehicle (BEV)   
3  MODEL S  Battery Electric Vehicle (BEV)   
4  MODEL Y  Battery Electric Vehicle (BEV)   

   Clean Alternative Fuel Vehicle (CAFV) Eligibility  Electric Range  \
0            Clean Alternative Fuel Vehicle Eligible             291   
1  Eligibility unknown as battery range has not b...               0   
2            Clean Alternative Fuel Vehicle Eligible             270   
3            Clean Alternative Fuel Vehicle Eligible             210   
4  Eligibility unknown as battery range has not b...               0   

   Base MSRP  Legislative District  DOL Vehicle ID  \
0          0                  37.0       125701579   
1          0                   1.0       244285107   
2          0                  36.0       156773144   
3          0                   5.0       165103011   
4          0                  23.0       205138552   

                Vehicle Location  \
0   POINT (-122.30839 47.610365)   
1  POINT (-122.179458 47.802589)   
2   POINT (-122.34848 47.632405)   
3   POINT (-122.03646 47.534065)   
4   POINT (-122.55717 47.733415)   

                                Electric Utility  2020 Census Tract  
0   CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA)       5.303301e+10  
1                         PUGET SOUND ENERGY INC       5.306105e+10  
2   CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA)       5.303301e+10  
3  PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)       5.303303e+10  
4                         PUGET SOUND ENERGY INC       5.303594e+10

### Creamos base de datos en PostgreSQL

```SQL
CREATE TABLE Electric_Vehicle_Population_Data(
    VIN_1_10 VARCHAR(20),
    County VARCHAR(50),
    City VARCHAR(50),
    State VARCHAR(2),
    Postal_Code INT,
    Model_Year SMALLINT,
    Make VARCHAR(50),
    Model VARCHAR(50),
    Electric_Vehicle_Type VARCHAR(50),
    Clean_Alternative_Fuel_Vehicle_CAFV_Eligibility VARCHAR(60),
    Electric_Range SMALLINT,
    Base_MSRP INT,
    Legislative_District SMALLINT,
    DOL_Vehicle_ID INT,
    Vehicle_Location VARCHAR(50),
    Electric_Utility VARCHAR(120),
    Census_Tract_2020 BIGINT
);
```

#### Se genera archivo de texto para leer la contraseña de la base de datos

In [3]:
ruta="D:/Study/Bootcamp Data Science BIT/Clases/Tutorías/Clase 12. 16-04-2024/"
with open(ruta+'password.txt','r') as f:
    pwd= f.read()

#### Se indican credenciales para crear conexión con la base de datos mediante sqlalchemy

In [4]:
from sqlalchemy import create_engine
hostname = 'localhost'
database = 'Project_DS_BIT'
username = 'postgres'
pwd = pwd
port_id = '5432'

#### Se establece conexión con la base de datos

In [5]:
engine = create_engine(f"postgresql://{username}:{pwd}@localhost:{port_id}/{database}")
engine

Engine(postgresql://postgres:***@localhost:5432/Project_DS_BIT)

#### Se indica estructura para obtener Dataframe de la base de datos

In [6]:
from sqlalchemy import text                                        
def runQuery(sql):                                                 
    result = engine.connect().execute((text(sql)))                 
    return pd.DataFrame(result.fetchall(), columns=result.keys())

#### Se lee Dataframe desde la base de datos PostgreSQL para gestionar y realizar consultas

In [7]:
datos = """
SELECT * FROM electric_vehicle_population_data
"""
runQuery(datos)

vin_1_10    county         city state  postal_code  model_year  \
0       1C4JJXR63P      King      Tukwila    WA      98188.0        2023   
1       1FTVW1EL3P      King  Federal Way    WA      98023.0        2023   
2       7SAYGDEE7P  Thurston      Olympia    WA      98506.0        2023   
3       5YJ3E1EC6N      King    Shoreline    WA      98133.0        2022   
4       1N4AZ0CPXE      King     Issaquah    WA      98027.0        2014   
...            ...       ...          ...   ...          ...         ...   
177861  WDDVP9AB4E      King     Bellevue    WA      98006.0        2014   
177862  KM8KNDAF2P      King      Seattle    WA      98115.0        2023   
177863  7SAYGDEE3P      King     Bellevue    WA      98005.0        2023   
177864  7SAYGDEE0P      King      Seattle    WA      98107.0        2023   
177865  1FTVW1EV1P      King    Shoreline    WA      98133.0        2023   

                 make     model                   electric_vehicle_type  \
0                JEEP  WRANGLER  Plug-in Hybrid Electric Vehicle (PHEV)   
1                FORD     F-150          Battery Electric Vehicle (BEV)   
2               TESLA   MODEL Y          Battery Electric Vehicle (BEV)   
3               TESLA   MODEL 3          Battery Electric Vehicle (BEV)   
4              NISSAN      LEAF          Battery Electric Vehicle (BEV)   
...               ...       ...                                     ...   
177861  MERCEDES-BENZ   B-CLASS          Battery Electric Vehicle (BEV)   
177862        HYUNDAI   IONIQ 5          Battery Electric Vehicle (BEV)   
177863          TESLA   MODEL Y          Battery Electric Vehicle (BEV)   
177864          TESLA   MODEL Y          Battery Electric Vehicle (BEV)   
177865           FORD     F-150          Battery Electric Vehicle (BEV)   

          clean_alternative_fuel_vehicle_cafv_eligibility  electric_range  \
0                   Not eligible due to low battery range              21   
1       Eligibility unknown as battery range has not b...               0   
2       Eligibility unknown as battery range has not b...               0   
3       Eligibility unknown as battery range has not b...               0   
4                 Clean Alternative Fuel Vehicle Eligible              84   
...                                                   ...             ...   
177861            Clean Alternative Fuel Vehicle Eligible              87   
177862  Eligibility unknown as battery range has not b...               0   
177863  Eligibility unknown as battery range has not b...               0   
177864  Eligibility unknown as battery range has not b...               0   
177865  Eligibility unknown as battery range has not b...               0   

        base_msrp  legislative_district  dol_vehicle_id  \
0               0                  11.0       238084297   
1               0                  30.0       259687921   
2               0                  22.0       251107874   
3               0                  32.0       209713906   
4               0                  41.0       113136768   
...           ...                   ...             ...   
177861          0                  41.0       121113408   
177862          0                  46.0       229604931   
177863          0                  41.0       236675442   
177864          0                  36.0       261447455   
177865          0                  32.0       258157553   

                       vehicle_location  \
0           POINT (-122.29179 47.43473)   
1           POINT (-122.36363 47.30675)   
2       POINT (-122.8874781 47.0519573)   
3           POINT (-122.34584 47.76726)   
4          POINT (-122.03646 47.534065)   
...                                 ...   
177861     POINT (-122.16937 47.571015)   
177862       POINT (-122.3185 47.67949)   
177863     POINT (-122.16085 47.624515)   
177864      POINT (-122.37815 47.66866)   
177865      POINT (-122.34584 47.76726)   

                                     electric

### Consultas sobre la base de datos

**1. ¿En qué año se registró la mayor cantidad de vehículos eléctricos?**

In [9]:
count_model = """
SELECT model_year, count(*) 
FROM electric_vehicle_population_data evpd 
GROUP BY model_year
ORDER BY count(*) DESC
"""
runQuery(count_model)

model_year  count
0         2023  57587
1         2022  27776
2         2021  19132
3         2018  14323
4         2020  11768
5         2019  10940
6         2017   8562
7         2024   7080
8         2016   5483
9         2015   4844
10        2013   4409
11        2014   3509
12        2012   1618
13        2011    775
14        2010     23
15        2008     20
16        2000      7
17        1999      5
18        2002      2
19        1998      1
20        2003      1
21        1997      1

**Respuesta**: Se evidencia que durante los últimos cuatro años, aumentó el registro de vehículos eléctricos cada año, respecto al anterior. En el año 2023 se evidenció la mayor cantidad de vehículos eléctricos registrados, aumentando más del doble la cantidad registrada el año anterior.

**Recomendación**: Para obtener la repuesta a la pregunta anterior, se recomienda tener en cuenta los datos de los últimos cinco años especialmente, debido a que fueron los años donde más se registraron vehículos eléctricos en el Estado de Washington en Estados Unidos.

**2. ¿Cuál es la marca de vehículos eléctricos que más se registró en los últimos cuatro años?**

In [10]:
count_make = """
SELECT make, model_year, count(*)
FROM electric_vehicle_population_data evpd 
GROUP BY make, model_year  
ORDER BY count(*) DESC
"""
runQuery(count_make)

make  model_year  count
0          TESLA        2023  28650
1          TESLA        2022  13751
2          TESLA        2021  11118
3          TESLA        2018   7915
4          TESLA        2020   7094
..           ...         ...    ...
247         FORD        1998      1
248       TOYOTA        2003      1
249  ROLLS ROYCE        2024      1
250      BENTLEY        2021      1
251   MITSUBISHI        2017      1

[252 rows x 3 columns]

**Respuesta:** Se evidencia que Tesla fue la marca de vehículos eléctricos que más se registró en los últimos cuatro años. 

**Recomendación**: Tener en cuenta que en el cálculo numérico, se evidencia el valor máximo de cantidad de vehículos y el año que sucedió por cada marca. 

**3. ¿Cuál es la ciudad donde se registraron la mayor cantidad de vehículos eléctricos?**

In [11]:
count_city = """
SELECT city, count(*) 
FROM electric_vehicle_population_data evpd 
GROUP BY city
ORDER BY count(*) DESC
LIMIT 20
"""
runQuery(count_city)

city  count
0         Seattle  29447
1        Bellevue   8931
2         Redmond   6478
3       Vancouver   6193
4         Bothell   5863
5        Kirkland   5362
6       Sammamish   5280
7          Renton   4980
8         Olympia   4299
9          Tacoma   3722
10     Bellingham   2941
11        Tukwila   2927
12           Kent   2782
13       Issaquah   2764
14        Spokane   2735
15       Lynnwood   2726
16        Everett   2475
17    Woodinville   2352
18  Mercer Island   2344
19      Snohomish   2102

**Respuesta:** Se evidencia que la ciudad de Seattle, es donde se han registrado la mayor cantidad de vehículos eléctricos del Estado de Washington, EEUU.

**Recomendación:** En el cálculo numérico, se tuvieron en cuenta todos los datos desde el primer año registrado y se podrá continuar obteniendo la información en el futuro.

**4. ¿Diferencia entre la cantidad de vehículos de batería eléctrica (BEV) y los Vehículos Eléctricos Híbridos (PHEVs)?**

In [12]:
count_evt = """
SELECT electric_vehicle_type, count(*)
FROM electric_vehicle_population_data evpd 
GROUP BY electric_vehicle_type
ORDER BY count(*) DESC
"""
runQuery(count_evt)

electric_vehicle_type   count
0          Battery Electric Vehicle (BEV)  139210
1  Plug-in Hybrid Electric Vehicle (PHEV)   38656

In [13]:
diferencia = """
SELECT MAX(count)-MIN(count) AS diferencia
FROM(
	SELECT electric_vehicle_type, count(*)
	FROM electric_vehicle_population_data evpd 
	GROUP BY electric_vehicle_type
	ORDER BY count(*) DESC
) AS COUNT_EVT
"""
runQuery(diferencia)

diferencia
0      100554

**Respuesta:** Se evidencia que los vehículos que tienen solo Batería Eléctrica son la mayoría con 139210 registros y triplican la cantidad de vehículos Híbridos con una diferencia de 100554 vehículos.

**Recomendación:** Para éste caso, todos los datos se tuvieron en cuenta para el cálculo. Además, está claro que los vehículos híbridos también son considerados en la categoría de vehículos eléctricos.

**5. ¿Cuál es el vehículo con mayor autonomía de utilización de su batería?**

In [14]:
autonomia = """
SELECT make, model, electric_range
FROM electric_vehicle_population_data evpd 
GROUP BY make, model, electric_range
ORDER BY electric_range DESC
LIMIT 30
"""
runQuery(autonomia)

make             model  electric_range
0       TESLA           MODEL S             337
1       TESLA           MODEL S             330
2       TESLA           MODEL 3             322
3       TESLA           MODEL 3             308
4       TESLA           MODEL X             293
5       TESLA           MODEL Y             291
6       TESLA           MODEL X             289
7       TESLA           MODEL S             270
8       TESLA           MODEL 3             266
9       TESLA           MODEL S             265
10  CHEVROLET           BOLT EV             259
11    HYUNDAI              KONA             258
12      TESLA           MODEL S             249
13      TESLA          ROADSTER             245
14        KIA              NIRO             239
15  CHEVROLET           BOLT EV             238
16      TESLA           MODEL X             238
17     JAGUAR            I-PACE             234
18   POLESTAR               PS2             233
19       AUDI            E-TRON             222
20      TESLA           MODEL 3             220
21      TESLA          ROADSTER             220
22       AUDI  E-TRON SPORTBACK             218
23     NISSAN              LEAF             215
24      TESLA           MODEL 3             215
25      TESLA           MODEL S             210
26      TESLA           MODEL S             208
27       AUDI            E-TRON             204
28    PORSCHE            TAYCAN             203
29      TESLA           MODEL X             200

**Respuesta:** En general, la marca Tesla tiene los vehículos eléctricos con mayor autonomía por millas, entre los cuales se evidencia el modelo S con 337 millas como el vehículo con mayor autonomía.

**Recomendación:** Tener en cuenta que existen diferentes vehículos con la misma marca y modelo, pero con diferente autonomía. 

**Se elimina engine para terminar la conexión con la base de datos**

In [15]:
del engine